File: src/main/java/ru/pgk/pgk/features/teacher/service/TeacherServiceImpl.java
package ru.pgk.pgk.features.teacher.service;

import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.pgk.pgk.features.department.entitites.DepartmentEntity;
import ru.pgk.pgk.features.teacher.dto.params.AddOrUpdateTeacherParams;
import ru.pgk.pgk.features.teacher.entities.TeacherEntity;
import ru.pgk.pgk.features.teacher.repositoties.TeacherRepository;
import ru.pgk.pgk.features.teacher.service.queries.TeacherQueriesService;

@Service
@RequiredArgsConstructor
public class TeacherServiceImpl implements TeacherService {

    private final TeacherRepository teacherRepository;

    private final TeacherQueriesService teacherQueriesService;

    private final EntityManager entityManager;

    @Override
    @Transactional
    @Caching(
            put = {
                    @CachePut(cacheNames = "TeacherQueriesService::getById", key = "#result.id"),
                    @CachePut(cacheNames = "TeacherQueriesService::getByCabinet", key = "#result.cabinet", condition = "#result.cabinet != null")
            },
            evict = {
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAllSearchByName", allEntries = true),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAll", allEntries = true)
            }
    )
    public TeacherEntity add(AddOrUpdateTeacherParams params) {
        TeacherEntity teacher = new TeacherEntity();
        setAddOrUpdateTeacherParams(teacher, params);
        return teacherRepository.save(teacher);
    }

    @Override
    @Transactional
    @Caching(
            put = {
                    @CachePut(cacheNames = "TeacherQueriesService::getById", key = "#result.id"),
                    @CachePut(cacheNames = "TeacherQueriesService::getByCabinet", key = "#result.cabinet", condition = "#result.cabinet != null")
            },
            evict = {
                    @CacheEvict(cacheNames = "ScheduleSearchService::getAllByTeacherId", key = "#id.toString() + '-' + '*'"),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAllSearchByName", allEntries = true),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAll", allEntries = true),
                    @CacheEvict(cacheNames = "ScheduleService::teacherGetByTeacherId", key = "'*' + '-' + #id")
            }
    )
    public TeacherEntity update(Integer id, AddOrUpdateTeacherParams params) {
        TeacherEntity teacher = teacherQueriesService.getById(id);
        setAddOrUpdateTeacherParams(teacher, params);
        return teacherRepository.save(teacher);
    }

    @Override
    @Transactional
    @Caching(
            evict = {
                    @CacheEvict(cacheNames = "TeacherQueriesService::getById", key = "#result.id"),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getByCabinet", key = "#result.cabinet", condition = "#result.cabinet != null "),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAllSearchByName", allEntries = true),
                    @CacheEvict(cacheNames = "TeacherQueriesService::getAll", allEntries = true),
                    @CacheEvict(cacheNames = "ScheduleSearchService::getAllByTeacherId", key = "#id.toString() + '-' + '*'"),
                    @CacheEvict(cacheNames = "ScheduleService::teacherGetByTeacherId", key = "'*' + '-' + #id")
            }
    )
    public TeacherEntity deleteById(Integer id) {
        TeacherEntity teacher = teacherQueriesService.getById(id);
        teacherRepository.delete(teacher);
        return teacher;
    }

    private void setAddOrUpdateTeacherParams(TeacherEntity teacher, AddOrUpdateTeacherParams params) {
        teacher.setFirstName(params.firstName());
        teacher.setLastName(params.lastName());
        teacher.setMiddleName(params.middleName());
        teacher.setCabinet(params.cabinet());
        teacher.setDepartments(params.departmentIds().stream().map(id -> entityManager.getReference(DepartmentEntity.class, id)).toList());
    }
}

File: teacher-service/src/main/kotlin/ru/pgk/teacher_service/features/services/TeacherServiceImpl.kt
package ru.pgk.teacher_service.features.services

import jakarta.persistence.EntityManager
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Caching
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.pgk.teacher_service.features.dto.params.AddOrUpdateTeacherParams
import ru.pgk.teacher_service.features.entities.DepartmentEntity
import ru.pgk.teacher_service.features.entities.TeacherEntity
import ru.pgk.teacher_service.features.repositories.TeacherRepository
import ru.pgk.teacher_service.features.services.queries.TeacherQueriesService

@Service
class TeacherServiceImpl(
    private val teacherRepository: TeacherRepository,
    private val teacherQueriesService: TeacherQueriesService,
    private val entityManager: EntityManager
) : TeacherService {

    @Transactional
    @Caching(
        put = [
            CachePut(
                cacheNames = ["TeacherQueriesService::getById"],
                key = "#result.id"
            ),
            CachePut(
                cacheNames = ["TeacherQueriesService::getByCabinet"],
                key = "#result.cabinet",
                condition = "#result.cabinet != null",
                unless = ""
            )
        ],
        evict = [
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAllSearchByName"],
                allEntries = true
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAll"],
                allEntries = true
            )
        ]
    )
    override fun add(params: AddOrUpdateTeacherParams): TeacherEntity {
        val teacher = TeacherEntity()
        setAddOrUpdateTeacherParams(teacher, params)
        return teacherRepository.save(teacher)
    }

    @Transactional
    @Caching(
        put = [
            CachePut(
                cacheNames = ["TeacherQueriesService::getById"],
                key = "#result.id"
            ),
            CachePut(
                cacheNames = ["TeacherQueriesService::getByCabinet"],
                key = "#result.cabinet",
                condition = "#result.cabinet != null"
            )
        ],
        evict = [
            CacheEvict(
                cacheNames = ["ScheduleSearchService::getAllByTeacherId"],
                key = "#id.toString() + '-' + '*'"
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAllSearchByName"],
                allEntries = true
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAll"],
                allEntries = true
            ),
            CacheEvict(
                cacheNames = ["ScheduleService::teacherGetByTeacherId"],
                key = "'*' + '-' + #id"
            )
        ]
    )
    override fun update(id: Int, params: AddOrUpdateTeacherParams): TeacherEntity {
        val teacher: TeacherEntity = teacherQueriesService.getById(id)
        setAddOrUpdateTeacherParams(teacher, params)
        return teacherRepository.save(teacher)
    }

    @Transactional
    @Caching(
        evict = [
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getById"],
                key = "#result.id"
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getByCabinet"],
                key = "#result.cabinet",
                condition = "#result.cabinet != null "
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAllSearchByName"],
                allEntries = true
            ),
            CacheEvict(
                cacheNames = ["TeacherQueriesService::getAll"],
                allEntries = true
            ),
            CacheEvict(
                cacheNames = ["ScheduleSearchService::getAllByTeacherId"],
                key = "#id.toString() + '-' + '*'"
            ),
            CacheEvict(
                cacheNames = ["ScheduleService::teacherGetByTeacherId"],
                key = "'*' + '-' + #id"
            )
        ]
    )
    override fun deleteById(id: Int): TeacherEntity {
        val teacher: TeacherEntity = teacherQueriesService.getById(id)
        teacherRepository.delete(teacher)
        return teacher
    }

    private fun setAddOrUpdateTeacherParams(teacher: TeacherEntity, params: AddOrUpdateTeacherParams) {
        teacher.firstName = params.firstName
        teacher.lastName = params.lastName
        teacher.middleName = params.middleName
        teacher.cabinet = params.cabinet
        teacher.departments = params.departmentIds.map { id ->
            entityManager.getReference(DepartmentEntity::class.java, id)
        }
    }
}